src/commands/dockerInstaller.ts

```diff
@@ -0,0 +1,100 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import ChildProcessProvider from '../debugging/coreclr/ChildProcessProvider';
+import { LocalFileSystemProvider } from '../debugging/coreclr/fsProvider';
+import { OSTempFileProvider } from '../debugging/coreclr/tempFileProvider';
+import { ext } from '../extensionVariables';
+import { localize } from '../localize';
+import { streamToFile } from '../utils/httpRequest';
+import LocalOSProvider from '../utils/LocalOSProvider';
+import { execAsync } from '../utils/spawnAsync';
+
+export abstract class DockerInstallerBase {
+    protected abstract downloadUrl: string;
+    protected abstract fileExtension: string;
+    protected abstract installationMessage: string;
+    protected abstract getInstallCommand(fileName: string): string;
+
+    public async downloadAndInstallDocker(): Promise<void> {
+        const confirmInstall: string = localize('vscode-docker.commands.DockerInstallerBase.confirm', 'Are you sure you want to install Docker on this machine?');
+        const installTitle: string = localize('vscode-docker.commands.DockerInstallerBase.install', 'Install');
+        const downloadingMessage: string = localize('vscode-docker.commands.DockerInstallerBase.downloading', 'Downloading Docker installer...');
+        let downloadedFileName: string;
+
+        // no need to check result - cancel will throw a UserCancelledError
+        await ext.ui.showWarningMessage(confirmInstall, { modal: true }, { title: installTitle });
+
+        await vscode.window.withProgress(
```

Comment:
It's not imperative but another way you can do this is like this:

```typescript
const downloadedFileName = await vscode.window.withProgress( { ... }, () => this.downloadInstaller() );
```

(There may need to be +/- some `async`/`await` in there, but in general `withProgress` can return the inner value.)